includes/forms/class-llms-forms.php

```diff
@@ -1002,6 +1019,37 @@ public function render_field_block( $html, $block ) {
 
 	}
 
+	/**
+	 * Maybe add the required email block to a form
+	 *
+	 * That's for checkout or registration forms for non logged in users.
+	 *
+	 * @since [version]
+	 *
+	 * @param array[] $blocks   Array of parsed WP_Block arrays.
+	 * @param string  $location The request form location ID.
+	 * @param array   $args     Additional arguments passed to the short-circuit filter.
+	 * @return array[]
+	 */
+	public function maybe_add_required_email_field_block( $blocks, $location, $args ) {
+
+		if ( is_user_logged_in() || ! in_array( $location, array( 'checkout', 'registration' ), true ) ) {
+			return $blocks;
+		}
+
+		foreach ( $this->get_field_blocks( $blocks ) as $field_block ) {
```

Comment:
We can use `LLMS_Forms::get_field_by( ... )` to pull the email field out of the form by its id and then check if the block is visible. If it returns false, add the field.

If it returns the field and the block isn't visible we should modify the visibility settings to make it visible instead of adding a new field dynamically. My thinking on this is that if someone has made customizations to the email address field (label, description, etc...) we want to ensure those still hold

Reply:
I see what you mean, yeah

Reply:
@thomasplevy 
what do you think about the invisible fields handling strategy in https://github.com/gocodebox/lifterlms/pull/1688/commits/7379f4ba582059c90ae560783edc6fbfc96752a7